src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp

```diff
@@ -5270,9 +5270,8 @@ void MacroAssembler::movoop(Register dst, jobject obj) {
     mov(dst, Address((address)obj, rspec));
   } else {
     address dummy = address(uintptr_t(pc()) & -wordSize); // A nearby aligned address
-    ldr_constant(dst, Address(dummy, rspec));
+    mov(dst, Address(dummy, rspec));
```

Comment:
- it is not a load from a Constant Pool, so calling ldr_constant here is seems incorrect
- the ldr_constant function utilizes either ldr (with a range limit of ±1MB) or, when -XX:-NearCpool is enabled, adrp (range limit of ±2GB) followed by ldr — both of which may fall short when mutable data is allocated on the C heap.

Reply:
This change looks wrong, for a number of reasons.  First, the dummy address would no longer be needed, and we could just use the same mov as the supports_instruction_patching() case.  However, if supports_instruction_patching() is false, I think we are not allowed to generate a multi-instruction movz/movk sequence.  We really need something like ldr_constant for this case, so that we load from memory.
However, as you point out, this is tied to NearCpool.  For a far metadata slot access, ADR+LDR is the right answer.  After this change, will there be any metadata left that could still benefit from NearCpool?   If not, then it might make sense to turn it off permanently.  Instead of choosing between PC-relative "ldr literal" and far ADR+LDR based on NearCpool, we could decide based on the distance to the metadata table. I believe "ldr literal" only has a 1MB range.
CC @theRealAph

Reply:
That's right. Thanks for pointing that out to me.
I have a fix for movoop issue on supports_instruction_patching=false case. Probably it should be considered as a separate change: https://github.com/openjdk/jdk/pull/22448